### PR TITLE
The EJS demo.

### PR DESCRIPTION
**Pull request by dkorolev**
Gentlemen,

Here's a relatively clean and simple way I have in mind to keep the "source of truth" of the "frontend" of the "app" on the Current side, while allowing to render it somewhat nicely for the end user, with one line change time on the frontend being, well, under one second.

Comments would be much appreciated.

Thanks,
Dima

### CODE REVIEW

File: examples/EJS/cpp/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+<h1>Primes from 99 to 101</h1>
+<ul>
+   <li>999</li> 
+</ul>
+```
+
+Once both servers are running, "HTML" "beautification" would work just like this:
+
+```
+$ curl -H "Accept: text/html" "localhost:3000?a=10&b=20"
```
Comment (sompylasar) [inline examples/EJS/cpp/README.md]: 3001?

File: examples/EJS/cpp/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+<h1>Primes from 10 to 20</h1>
+<ul>
+   <li>11</li>  <li>13</li>  <li>17</li>  <li>19</li> 
+</ul>
+```
+
+And if that's too nerdy, just open [http://localhost:3000/?a=10&b=20] from the browser.
```
Comment (sompylasar) [inline examples/EJS/cpp/README.md]: 3001?

File: examples/EJS/cpp/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# Current + EJS Demo
+
+Just `make` in this directory would build and start a simple server on port 3000:
```
Comment (sompylasar) [inline examples/EJS/cpp/README.md]: Which port serves what? Not obvious, confusing.

File: examples/EJS/nodejs/package.json [hunk 4]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "CurrentEJS",
```
Comment (sompylasar) [inline examples/EJS/nodejs/package.json]: Packages are generally lowercase and dash-separated. So, `current-ejs` or `current-demo-renderer-ejs` or `c5t-demo-renderer-ejs`.

File: examples/EJS/nodejs/package.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  "keywords": [
+    "current",
+    "ejs",
+    "cpp",
+    "backend"
+  ],
+  "main": "server.js",
+  "dependencies": {
+    "body-parser": "*",
+    "express": "*",
```
Comment (sompylasar) [inline examples/EJS/nodejs/package.json]: I'd rather not use wildcard versions, stick to the current major at least.

File: examples/EJS/nodejs/template.ejs [hunk 6]
```diff
@@ -0,0 +1,4 @@
+<h1>Primes from <%= a %> to <%= b %></h1>
+<ul>
+  <% for(var i = 0; i < x.length; ++i) { %> <li><%= x[i] %></li> <% } %>
```
Comment (sompylasar) [inline examples/EJS/nodejs/template.ejs]: You don't need whitespace between `%>` and `<`. Whitespace is sometimes meaningful in HTML (if the blocks are inline, the space gets a nonzero width).

Comment (sompylasar) [inline examples/EJS/nodejs/template.ejs]: And `> <% ` of course.

File: examples/EJS/cpp/ejs.cc [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+              }
+            }
+            if (is_prime) {
+              response.x.push_back(p);
+            }
+          }
+          return response;
+        }();
+
+        const bool html = [&request]() {
```
Comment (sompylasar) [inline examples/EJS/cpp/ejs.cc]: Why the calculating server knows about the renderer(s)?

File: examples/EJS/cpp/README.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+```
+$ curl "localhost:3000?a=10&b=20"
+{"a":10,"b":20,"x":[11,13,17,19]}
+```
+
+Adding `npm run start` from `../nodejs` would start another, node.js, rendering server on port 3001:
```
Comment (mzhurovich) [inline examples/EJS/cpp/README.md]: `npm install` first?

File: examples/EJS/cpp/README.md (hunk 3 above)
Comment (dkorolev) [inline examples/EJS/cpp/README.md]: Updated README in 83cd181.

File: examples/EJS/cpp/README.md (hunk 1 above)
Comment (dkorolev) [inline examples/EJS/cpp/README.md]: It's how it should be. Updated README in 83cd181.

File: examples/EJS/cpp/README.md (hunk 2 above)
Comment (dkorolev) [inline examples/EJS/cpp/README.md]: It's how it should be. Updated README in 83cd181.

File: examples/EJS/cpp/README.md (hunk 8 above)
Comment (dkorolev) [inline examples/EJS/cpp/README.md]: :+1: b8c5de8

File: examples/EJS/cpp/ejs.cc (hunk 7 above)
Comment (dkorolev) [inline examples/EJS/cpp/ejs.cc]: It has the `--ejs_renderer_port` flag set, which defaults to `3001`. If `Accept: text/html` is in the headers, the rendered is invoked.

File: examples/EJS/nodejs/package.json (hunk 5 above)
Comment (dkorolev) [inline examples/EJS/nodejs/package.json]: :+1: d1f3a14

File: examples/EJS/nodejs/template.ejs (hunk 6 above)
Comment (dkorolev) [inline examples/EJS/nodejs/template.ejs]: :+1: 0136832

File: examples/EJS/nodejs/package.json (hunk 4 above)
Comment (dkorolev) [inline examples/EJS/nodejs/package.json]: :+1:

**Comment by dkorolev** [comment]
@sompylasar @grixa @mzhurovich PTAL && thanks!

File: examples/EJS/README.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+## Details
+
+Just `make` in the `EJS/cpp` directory would build and start a simple server on port 3000:
+
+```
+$ curl "http://localhost:3000?a=10&b=20"
+{"a":10,"b":20,"x":[11,13,17,19]}
+```
+
+Indepenently from the C++ binary, running `npm i && npm run start` from `EJS/nodejs` would start another, node.js, rendering server on port 3001:
```
Comment (sompylasar) [inline examples/EJS/README.md]: Indepen*d*ently

File: examples/EJS/nodejs/package.json [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  ],
+  "keywords": [
+    "current",
+    "ejs",
+    "cpp",
+    "backend"
+  ],
+  "main": "server.js",
+  "dependencies": {
+    "body-parser": ">=1.15.2",
```
Comment (sompylasar) [inline examples/EJS/nodejs/package.json]: Why `>=`? Either nothing (exact version), or `^` (latest with the same major version).

**Comment by dkorolev** [comment]
@sompylasar :+1: :+1: :+1: 70825ef

**Comment by mzhurovich** [comment]
The other possible design is to make Node.JS app a "rendering layer" that forwards requests to the Current binary, which in its turn serves only data JSONs and knows nothing about any renders etc. (what @sompylasar said). This example stays as is for now.
CC @grixa

**Pull request merged**